src/Dscp/Educator/Validate.hs

```diff
@@ -13,63 +14,89 @@ import Data.Text.Buildable (build)
 import Dscp.Crypto (Hash, PublicKey, MerkleSignature, fromFoldable, getMerkleRoot, hash, verify)
-import Dscp.Educator.Block (PrivateBlock (..), PrivateBlockHeader (..), pbBody, pbbTxs)
+import Dscp.Educator.Block (PrivateBlock (..), PrivateBlockHeader (..), _pbBody, _pbbTxs)
 import Dscp.Educator.Serialise ()
 import Dscp.Educator.Txs (PrivateTx (..))
 
 import qualified Text.Show
 import qualified Data.Text.Buildable ()
 
 
--- | Validation failures
-data ValidationFailure
-    = MerkleSignatureMismatch !(MerkleSignature PrivateTx) !(MerkleSignature PrivateTx)
-      -- ^ Header merkle tree root does not match
-      -- body merkle tree root
-    | TxPublicKeyMismatch !(Hash PublicKey) !(Hash PublicKey)
-      -- ^ Witness public key do not match transaction address
-    | TxSignatureMismatch !PublicKey !SubmissionSig !(Hash Submission)
-      -- ^ Witness signature do not match transaction submission signature
+-- | Block validation failures
+data BlockValidationFailure
+    = -- | Header merkle tree root does not match body merkle tree root
+      MerkleSignatureMismatch
+        { bvfExpectedSig     :: !(MerkleSignature PrivateTx)
+        , bvfActualMerkleSig :: !(MerkleSignature PrivateTx) }
+      -- | Submission failure
+    | SubmissionInvalid
+        { bvfSubmissionFailure :: !SubmissionValidationFailure }
     deriving (Eq)
 
-instance Show ValidationFailure where
-    show e = toString . pretty $ e
+-- | Submission validation failures
+data SubmissionValidationFailure
+    = -- | Witness public key do not match submission address
+      SubmissionPublicKeyMismatch
+        { svfExpectedPubKey :: !(Hash PublicKey)
+        , svfActualPubKey   :: !(Hash PublicKey) }
+      -- | Witness signature do not match submission signature
+    | SubmissionSignatureMismatch
+        { svfSubmissionHash   :: !(Hash Submission)
+        , svfSubmissionSigKey :: !PublicKey
+        , svfSubmissionSig    :: !SubmissionSig }
+    deriving (Eq)
+
+instance Show BlockValidationFailure where
+    show = toString . pretty
 
-instance Buildable ValidationFailure where
-    build (MerkleSignatureMismatch expected got) =
-      "Merkle tree root signature mismatch. Expected " `mappend` show expected
-      `mappend` " got " `mappend` show got
-    build (TxPublicKeyMismatch expected got) =
-      "Tx public key address mismatch. Expected " `mappend` show expected `mappend`
-      " got " `mappend` show got
-    build (TxSignatureMismatch key sig h) =
-      "Tx signature mismatch. Submission data " `mappend` show h `mappend` " with signature "
-      `mappend` show sig `mappend` " does not correspond to public key " `mappend` show key
+instance Buildable BlockValidationFailure where
+    build (MerkleSignatureMismatch {..}) =
+      "Merkle tree root signature mismatch. Expected " `mappend` show bvfExpectedSig
+      `mappend` " got " `mappend` show bvfActualMerkleSig
+    build (SubmissionInvalid (SubmissionPublicKeyMismatch {..})) =
```

Comment:
There should be a separate `Buildable` instance for `SubmissionValidationFailure` instead, and case for `SubmissionInvalid` should use this instance.